### ISSUE TITLE
Incorrect type for `FileWithPath`'s `path` and `relativePath` properties

### ISSUE THREAD

**Issue by bthall16** [react-dropzone/file-selector#115]
The `FileWithPath` interface defines both its `path` and `relativePath` properties as optional:

https://github.com/react-dropzone/file-selector/blob/f159a4a034b942054ad18453fffa77fce7a6dc87/src/file.ts#L1233-L1237

indicating they _could_ be `undefined`.

Looking at the implementation of `toFileWithPath`, however, it looks like there's no code path where _either_ could be `undefined`. In particular, [if `file.path` isn't already a string, it's set to a variable which is definitely a string by that point](https://github.com/react-dropzone/file-selector/blob/f159a4a034b942054ad18453fffa77fce7a6dc87/src/file.ts#L1217-L1219). [`relativePath` is unconditionally set to the same variable mentioned above that's definitely a string](https://github.com/react-dropzone/file-selector/blob/f159a4a034b942054ad18453fffa77fce7a6dc87/src/file.ts#L1228-L1229).

This seems like a bug with the definition of `FileWithPath`: if neither `path` nor `relativePath` can ever be `undefined`, then marking them as optional isn't accurate.

I came across this as I was trying to figure out which situations could cause `path` or `relativePath` to be `undefined` so I could account for those situations in my web app.

**Comment by jonkoops** [comment]
I'm working on improving the correctness of the internal types as we still have to many places where things are `any` (implicit or otherwise). So I will see if I can resolve this as part of that work.

**Comment by bthall16** [comment]
Thank you for the reply!

**Comment by rolandjitsu** [comment]
Probably the reason it is that way is because the type is passed around as input too:

https://github.com/react-dropzone/file-selector/blob/f159a4a034b942054ad18453fffa77fce7a6dc87/src/file.ts#L1206

At that point, the `path` could be undefined.

I guess splitting up the input from the output may make sense. Just not sure if that'll break downstream packages w/o making a breaking change.

**Comment by jonkoops** [comment]
> I guess splitting up the input from the output may make sense. Just not sure if that'll break downstream packages w/o making a breaking change.

Good point, we'll have to investigate this. We're doing a major bump for what is on the `beta` branch anyways, so that might be a good time to do this if we want to.

**Comment by bthall16** [comment]
Making `path` non-optional could be a breaking change for someone using `FileWithPath` as an input type. 

I think* it wouldn't be breaking to have a separate interface where `path` is required _if_ that interface is only used as part of the return type for `fromEvent` and friends: any existing code written under the assumption `path` could be `undefined | string` would still work if `path` was only ever `string`.

\* wouldn't be surprised if there are cases in the wild where this doesn't hold

**Comment by jonkoops** [comment]
Agreed. We're bumping a major anyways, so I would like to ride that and make a clean cut. I plan to do a bunch of refactoring and cleanup to familiarize myself with the code base fully before making this change, so we'll see when I get to it.

**Comment by alexandre-combemorel** [comment]
I'm using react-dropzone to upload element with aws s3 protocol and it seems that it trigger an issue of signature on the back end side.
I narrowed done the issue to this line: ([here](https://github.com/react-dropzone/file-selector/blob/f159a4a034b942054ad18453fffa77fce7a6dc87/src/file.ts#L1216))
```ts
: `./${file.name}`;
```
while it used to be 
```ts
: file.name;
```
I don't understand why we are updating the path to a string with relative path?

If I change it back to the actual original name, it works without issue.

**Comment by lomonosv** [comment]
I have the same issue with `./` at the beginning of the file name.

**Comment by rolandjitsu** [comment]
@alexandre-combemorel 

> I don't understand why we are updating the path to a string with relative path?

because we treat it as a relative path and not having the dot is ambiguous . You could switch to the file name, if that's what you're interested in. 

But I see how this is causing some inconvenience.
